### PR TITLE
Feature ticket page rework

### LINKED ISSUE #230: Open order is not loaded

### LINKED ISSUE #257: Can't accept terms & conditions on portrait screens.

### LINKED ISSUE #272: Account details intercept flow
Currently, we check the profile details before the order creation. This needs to be done on checkout. Preferably in a dialog before the T&C.

### LINKED ISSUE #344: Refactor lancie-ticket-block
During the AV, someone suggested for the Early bird tickets to not have a set number, but a time limit instead. After our meeting, we decided that we want that too. I checked throughout our system, and the implementation is already there, except in this element early birds cannot have a deadline. Refactoring this will also help removing the hardcoded tickettypes, and moving to dynamic tickets.

The new code should check if either the deadline, the limit, or both are set, and handle accordingly.

### PR DESCRIPTION
**Pull request by martijnjanssen**
Requires https://github.com/AreaFiftyLAN/lancie-api/pull/314

Fixes #344 
Fixes #272 
Fixes #257 
Fixes #230 

~~So right now it is not working as intended. In `lancie-login-check` the `this.$` selector seems to be broken for no reason. Could you look at this @TimvdLippe ?~~

### CODE REVIEW

File: src/lancie-login/lancie-login-check.html [hunk 1]
```diff
@@ -40,6 +44,18 @@
         checks.push(function(newUser) {
           this.user = newUser;
         }.bind(this));
+
+        // Add eventlistener to catch logged-in event in entire element
+        this.addEventListener('logged-in', (e) => {
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-check.html]: Change to `function(e) {` ES5 syntax

File: src/lancie-ticket-page/lancie-ticket-basket.html [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+<link rel="import" href="../lancie-login/lancie-ajax.html">
+<link rel="import" href="lancie-ticket-basket-item.html">
+
+<dom-module id="lancie-ticket-basket">
+  <template>
+	<style>
+	</style>
+
+    <lancie-ajax id="removeTicket" refurl="/orders/[[basket.id]]" method="DELETE" on-response="onRemoveTicket"></lancie-ajax>
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-basket.html]: It seems that the indentation is off in this file.

Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-basket.html]: There seems to be tab characters in this file, instead of 2 spaces

**Comment by MatthijsKok** [comment]
#344 can nicely be done as well.

**Comment by martijnjanssen** [comment]
@TimvdLippe could you look at this?

**Comment by skamoen** [comment]
Clicktesting issues 1

1. Create an order without logging in
2. Login with existing user
3. Go through mollie
4. Callback to `/ordersuccess?order=...` results in an error page

The X-Auth-Token header is empty, therefore the orderstatus call results in 403

**Comment by skamoen** [comment]
Clicktesting issue 2

1. Create an order without loggin in
2. Register a new user
3. Click the link in the mail

If there's an orderId in the url, you can assume the user wants to finish the order.

**Comment by skamoen** [comment]
Clicktesting issue 3

Profile intercept form:
- 409 Profile Name in use error isn't handled, shows as unable to connect in the background. You want these errors more visible
- Succesful submit doesn't close the dialog. Dialog has to be closed and needs another click on checkout before I can continue.

File: src/lancie-login/lancie-login-form.html [hunk 3]
```diff
@@ -128,8 +128,11 @@
         }
       },
 
-      tryRegister: function() {
+      tryRegister: function(id) {
         if (this.validateRegisterForm()) {
+          if (!!id) {
+            this.$.ajaxRegister.body['orderId'] = id;
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-form.html]: This feels very wrong. Maybe expose the `body` property so that users of the login-form can specify which body to send?

File: src/lancie-login/lancie-login-form.html [hunk 4]
```diff
@@ -170,12 +173,11 @@
...
         });
-     },
+      },
 
-      onRegister: function(e, response) {
-        if (response.succeeded) {
-          this.fire('navigate', {page: 'check-email'});
-        }
+      onRegister: function(e, request) {
+        this.fire('navigate', {page: 'check-email'});
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-form.html]: The condition for `response.succeeded` has been removed. Intended?

File: src/lancie-login/lancie-login-form.html [hunk 5]
```diff
@@ -191,7 +193,7 @@
       bodyFrom: function(username, password) {
         return {
           username: username,
-          password: password
+          password: password,
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-form.html]: Extra comma is unnecessary, but does not harm.

File: src/lancie-login/lancie-register-confirmation.html [hunk 6]
```diff
@@ -44,26 +44,25 @@
...
       },
 
       attached: function() {
         this.confirming = true;
         this.success = false;
         this.error = false;
         this.tokenTwice = false;
-        this.token = window.location.href.split('token=')[1];
-        this.tryRequestPasswordReset();
+        this.search = window.location.search;
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-register-confirmation.html]: This can change over time. You can also use `<iron-location>` with the `query` property: https://beta.webcomponents.org/element/PolymerElements/iron-location/iron-location#property-query

File: src/lancie-login/lancie-register-confirmation.html [hunk 7]
```diff
@@ -75,7 +74,7 @@
             this.tokenTwice = true;
           }
         }
-      }
+      },
```
Comment (TimvdLippe) [inline src/lancie-login/lancie-register-confirmation.html]: Not explicitly required.

File: src/lancie-my-area/profile-edit-form.html [hunk 8]
```diff
@@ -1,4 +1,9 @@
 <link rel="import" href="../../bower_components/iron-flex-layout/iron-flex-layout-classes.html">
+<link rel="import" href="../../bower_components/paper-item/paper-icon-item.html">
+<link rel="import" href="../../bower_components/paper-dropdown-menu/paper-dropdown-menu.html">
+<link rel="import" href="../../bower_components/iron-icons/communication-icons.html">
+<link rel="import" href="../../bower_components/iron-icons/social-icons.html">
+<link rel="import" href="../../bower_components/paper-input/paper-input.html">
```
Comment (TimvdLippe) [inline src/lancie-my-area/profile-edit-form.html]: :+1: :+1: :+1: :+1: :+1: :+1:

File: src/lancie-my-area/profile-edit-form.html [hunk 9]
```diff
@@ -53,7 +58,6 @@
         <paper-input label="Phone Number" type="text" value="{{profile.phoneNumber}}"></paper-input>
       </div>
     </div>
-    <paper-icon-item>
```
Comment (TimvdLippe) [inline src/lancie-my-area/profile-edit-form.html]: :joy: Oops

File: src/lancie-ticket-page/lancie-profile-dialog.html [hunk 10]
```diff
@@ -49,24 +45,24 @@
 
     <lancie-ajax id="ajaxUpdateProfile" refurl="users/current/profile" method="POST" on-response="onProfileUpdate"></lancie-ajax>
 
-    <lancie-card heading="Fill in your profile">
+    <paper-dialog id="dialog" entry-animation="scale-up-animation" exit-animation="fade-out-animation" with-backdrop>
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-profile-dialog.html]: Shouldn't this be `lancie-dialog`?

File: src/lancie-ticket-page/lancie-profile-dialog.html [hunk 11]
```diff
@@ -80,14 +76,14 @@
           type: Boolean,
           value: true
         },
-        route: {
-          type: String,
-          notify: true
-        }
+      },
+
+      ready: function() {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-profile-dialog.html]: `ready` => `attached`?

Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-profile-dialog.html]: Also on `detached` we probably need to remove the dialog again.

File: src/lancie-ticket-page/lancie-ticket-basket-item.html [hunk 12]
```diff
@@ -61,10 +60,10 @@
       removeTicket: function() {
         this.fire('remove-ticket', {
           type: this.ticket.type.name,
-          options: this.ticket.enabledOptions.map(function(option) {
+          options: this.ticket.enabledOptions.map(function       (option) {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-basket-item.html]: Unnecessary whitespace

File: src/lancie-ticket-page/lancie-ticket-page.html [hunk 13]
```diff
@@ -31,102 +32,120 @@
...
+            value: [],
+          },
         },
-        basket: {
-          type: Object,
-          value: {},
-          // notify: true,
-          observer: '_basketChanged',
+
+        ready: function () {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-page.html]: `attached`

File: src/lancie-ticket-page/lancie-ticket-page.html [hunk 14]
```diff
@@ -31,102 +32,120 @@
...
+          }
+        },
+
+        onCoupleOrder: function (e, request) {
+          if (request.succeeded) {
+            this.basket = request.response.object;
+            this.$.orderStorage.removeStorage();
+          } else {
+            this.basket.status = 'ANONYMOUS';
+            this.fire('toast', {text: 'Unable to couple the order to your account. Try refreshing the page.'});
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-page.html]: Instead of refreshing, can we maybe fix the status?

**Comment by TimvdLippe** [comment]
I have updated the PR description to correctly reference multiple issues to be fixed

**Comment by skamoen** [comment]
clicktesting issue:

When coming from the confirm email (with orderId in url), the confirmRegistrationpage should say "Click here to continue your order". That link goes to the `/tickets` page, but should also automatically trigger the popup window, for example with `/tickets?login`

**Comment by TimvdLippe** [comment]
Opening https://127.0.0.1:5100/tickets and logging throws a 401 error. Subsequently adding tickets to the order shows the toast "Unabled to connect to the server". Refreshing the page fixes this issue.

**Comment by TimvdLippe** [comment]
Adding a ticket with no ticket options selected and a ticket with all options selected, deleting the ticket with no options selected sometimes removes the other ticket.

File: src/lancie-content.html [hunk 15]
```diff
@@ -166,7 +166,7 @@
       },
 
       _filterConstruction: function(page) {
-        if (window.location.host.indexOf('localhost') !== -1) {
+        if (window.location.host.indexOf('localhost') >= 0 || window.location.host.indexOf('127.0.0.1') >= 0) {
```
Comment (martijnjanssen) [inline src/lancie-content.html]: Shouldn't this be `> -1`?

Comment (TimvdLippe) [inline src/lancie-content.html]: Logic is equivalent. You can do so, doesn't really matter I think

File: src/lancie-login/lancie-login-form.html (hunk 4 above)
Comment (martijnjanssen) [inline src/lancie-login/lancie-login-form.html]: This is done because this function is called on the `response` event. If the request was not succeeded, the `error` event would be fired, correct?

File: src/lancie-ticket-page/lancie-profile-dialog.html (hunk 10 above)
Comment (martijnjanssen) [inline src/lancie-ticket-page/lancie-profile-dialog.html]: It could be, but I am still hesitant to do so, I am still not quite sure what is going to happen to the current `lancie-dialog` element. Are we going to update the element or are we going to build a new `lancie-dialog`?

File: src/lancie-login/lancie-login-form.html (hunk 3 above)
Comment (martijnjanssen) [inline src/lancie-login/lancie-login-form.html]: So outside of the element you retrieve some data from the element and then set the body with the data you retrieved from the element?

**Comment by martijnjanssen** [comment]
Did you mean 'logging in' in [this](https://github.com/AreaFiftyLAN/lancie-frontend/pull/323#issuecomment-269988403) comment?

File: src/lancie-login/lancie-login-form.html (hunk 3 above)
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-form.html]: The user of lancie-login-form creates the body, not the form itself.

File: src/lancie-login/lancie-login-form.html (hunk 4 above)
Comment (TimvdLippe) [inline src/lancie-login/lancie-login-form.html]: Not sure. Probably is but not 100% sure

**Comment by martijnjanssen** [comment]
I think this is the final version of the new ticket page. @TimvdLippe what do you think of the fixes?

I have some doubts about the extraction of the methods in login-form. I would like to change the login form to a iron-form soon. But this will not be possible when the methods are outside of the form itself. For instance submit on enter will not work in that case.

**Comment by MatthijsKok** [comment]
Email check happens after page is unloaded.
- Make sure you're not logged in.
- Click on My Area.
- Click on Seatmap.
- Click on My Area again.
- Observe that the validation happened, without ever typing something.

**Comment by martijnjanssen** [comment]
You are correct, this is a thing that happens, but this is due to the fact that chrome has autofocus on the username field. If you observe what happens you see that the field is focussed on seeing the username field. When changing the page to another one, focus is lost. This is called the 'blur' event, which triggers validation of the field. If you want to change anything about this, please open an issue, or even better, a PR. But again, this is not an issue introduced by this PR. This won't be fixed here.

File: src/lancie-ticket-page/lancie-ticket-item.html [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+        this.options = options;
+        this.available = this._available();
+      },
+
+      addToBasket: function() {
+        if (!this.basket.tickets) {
+          this.$.addTicket.refurl = 'orders';
+        } else if (this.basket.tickets.length >= 5) {
+          this.fire('toast', {text: 'Maximum of 5 tickets.'});
```
Comment (MatthijsKok) [inline src/lancie-ticket-page/lancie-ticket-item.html]: Would like this changed to: "A maximum of 5 tickets can be added."

Comment (MatthijsKok) [inline src/lancie-ticket-page/lancie-ticket-item.html]: "Maximum of 5 tickets in basket." is even better

File: src/lancie-ticket-page/lancie-ticket-basket.html [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+  (function() {
+	Polymer({
+	  is: 'lancie-ticket-basket',
+	  properties: {
+      user: Object,
+		  basket: {
+			  type: Object,
+			  notify: true,
+		  },
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-basket.html]: Indentation is off here

File: src/lancie-ticket-page/lancie-ticket-basket.html [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+      if (!this.$.profileDialog.allFilledIn(this.user)) {
+        this.$.profileDialog.open();
+      } else {
+        this.$.termsOfService.open();
+      }
+    },
+
+    onCheckout: function(e, request) {
+      if (request.succeeded) {
+        window.location.href = request.response.object;
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-basket.html]: Add a comment this redirects to Mollie, so it is not an internal url change

File: src/lancie-ticket-page/lancie-ticket-item.html [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+        },
+        color: String,
+      },
+      observers: [
+        '_calulatePrice(options.*)'
+      ],
+
+      ready: function() {
+        var options = [];
+        for (var option in this.ticket.possibleOptions) {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-item.html]: Ticket might be undefined here? Better to add it in an observer

File: src/lancie-ticket-page/lancie-ticket-item.html [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+          return 'Available until: ' + new Date(this.ticket.deadline).toLocaleString('nl', {hour12: false});
+        } else if (deadlineType === 'limit') {
+          return 'Only ' + (ticket.limit - ticket.numberSold) + ' left!';
+        } else {
+          return 'Available';
+        }
+      },
+
+      _calulatePrice: function(changed) {
+        if (!this.ticket) {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-item.html]: Add this as argument to the observer, makes sure ticket is never undefined

File: src/lancie-ticket-page/lancie-ticket-item.html [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+        classes += (available) ? '' : ' sold-out';
+        return classes;
+      },
+
+      _hasOptions: function(options) {
+        return options.length > 0;
+      },
+
+      _available: function() {
+        var deadlineType = this.deadlineDisplay[this.ticket.ticketType];
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-item.html]: Also has a dependency on ticket

File: src/lancie-ticket-page/lancie-ticket-page.html [hunk 22]
```diff
@@ -4,263 +4,154 @@
...
+              this.fire('toast', {text: 'Unable to couple the order to your account. Try refreshing the page.'});
+            }
+          }
+        },
+
+        /*
+          If there is an order and the user logs in later, couple the order to the user.
+         */
+        _userChanged: function (user) {
+          if (this.basket.status === 'ANONYMOUS' && !!user) {
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-ticket-page.html]: Dependency on basket

File: src/lancie-ticket-page/lancie-tickets.html [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+              'green',
+              'purple',
+              'orange',
+            ];
+          },
+        },
+      },
+
+      _getColorClass: function(index) {
+        return this.colors[index];
```
Comment (TimvdLippe) [inline src/lancie-ticket-page/lancie-tickets.html]: What if index is greater than the length?
